Userland/Applications/Piano/AudioPlayerLoop.cpp

```diff
@@ -6,18 +6,94 @@
  */
 
 #include "AudioPlayerLoop.h"
-
+#include "Music.h"
 #include "TrackManager.h"
+#include <AK/Assertions.h>
 #include <AK/FixedArray.h>
+#include <AK/Forward.h>
+#include <AK/NonnullOwnPtr.h>
 #include <AK/NumericLimits.h>
+#include <AK/StdLibExtras.h>
 #include <LibAudio/ConnectionToServer.h>
+#include <LibAudio/Queue.h>
 #include <LibAudio/Resampler.h>
 #include <LibAudio/Sample.h>
-#include <LibCore/EventLoop.h>
+#include <LibIPC/Connection.h>
+#include <LibThreading/Thread.h>
+#include <sched.h>
+#include <time.h>
+
+struct AudioLoopResponsivenessTimer final : public IPC::ResponsivenessTimer {
+    AudioLoopResponsivenessTimer(IPC::ConnectionBase& parent)
+        : IPC::ResponsivenessTimer(parent)
+    {
+        update_time();
+    }
+
+    void update_time()
+    {
+        clock_gettime(CLOCK_REALTIME, &last_check_time);
+        if (!is_running)
+            return;
+
+        if (ns_duration_between(start_time, last_check_time) >= IPC::RESPONSIVENESS_TIMEOUT_MS * 1'000'000LL) {
+            timer_expired();
+            stop();
+        }
+    }
+
+    static long ns_duration_between(timespec start, timespec end)
+    {
+        auto s_duration = static_cast<i64>(end.tv_sec - start.tv_sec);
```

Comment:
Shold we add a full second for when `end.tv_nsec - start.tv_nsec < 0` ?

```cpp
  if ((stop->tv_nsec - start->tv_nsec) < 0) {
        result->tv_sec = stop->tv_sec - start->tv_sec - 1;
        result->tv_nsec = stop->tv_nsec - start->tv_nsec + 1000000000;
    } else {
        result->tv_sec = stop->tv_sec - start->tv_sec;
        result->tv_nsec = stop->tv_nsec - start->tv_nsec;
    }
    ```

Reply:
Or... we could just use AK/Time.h and its AK::Time class, and all the fun helpers in that class, e.g. 

<https://github.com/SerenityOS/serenity/blob/master/AK/Time.h#L270-L279>

Reply:
I totally missed that AK::Time exists.